Source/Libraries/GSF.Security/OIDCSecurityProvider.cs

```diff
@@ -0,0 +1,582 @@
+        }
+
+        /// <summary>
+        /// The Claim used to get the Roles for the user
+        /// </summary>
+        public string RolesClaim
+        {
+            get;
+            set;
+        }
+
+        #endregion
+
+        #region [ Methods ]
+
+        /// <summary>
+        /// Loads saved security provider settings from the config file if the <see cref="SecurityProviderBase.PersistSettings"/> property is set to true.
+        /// </summary>
+        /// <exception cref="ConfigurationErrorsException"><see cref="SecurityProviderBase.SettingsCategory"/> has a value of null or empty string.</exception>
+        public override void LoadSettings()
+        {
+            base.LoadSettings();
+
+            // Make sure default settings exist
+            ConfigurationFile config = ConfigurationFile.Current;
+            CategorizedSettingsElementCollection settings = config.Settings[SettingsCategory];
+
+            settings.Add(nameof(ClientID), "", "Defines the ClientID as required per OpenID Connect Standard.");
+            settings.Add(nameof(Scope), "", "Defines the Scope as required per OpenID Connect Standard.");
+            settings.Add(nameof(AuthorizationEndpoint), "", "Defines the Endpoint to redirect the user for Authorization.");
+            settings.Add(nameof(RedirectURI), "", "Defines the URI the User get's redirected to after signing in.");
+            settings.Add(nameof(ClientSecret), "", "Defines the Client Secret as required per OpenID Connect Standard for decrypting User Information.");
+            settings.Add(nameof(TokenEndpoint), "", "Defines the Endpoint to use to grab the User Token.");
+            settings.Add(nameof(RolesClaim), "roles", "Defines the claim used to identify the users roles.");
+            
+            ClientID = settings[nameof(ClientID)].ValueAs(ClientID);
+            Scope = settings[nameof(Scope)].ValueAs(Scope);
+            AuthorizationEndpoint = settings[nameof(AuthorizationEndpoint)].ValueAs(AuthorizationEndpoint);
+            RedirectURI = settings[nameof(RedirectURI)].ValueAs(RedirectURI);
+            ClientSecret = settings[nameof(ClientSecret)].ValueAs(ClientSecret);
+            TokenEndpoint = settings[nameof(TokenEndpoint)].ValueAs(TokenEndpoint);
+            RolesClaim = settings[nameof(RolesClaim)].ValueAs(RolesClaim);
+        }
+        
+        /// <summary>
+        /// Refreshes the <see cref="UserData"/>.
+        /// </summary>
+        /// <returns>true if <see cref="SecurityProviderBase.UserData"/> is refreshed, otherwise false.</returns>
+        public override bool RefreshData()
+        {
+            if (string.IsNullOrEmpty(UserData.Username))
+                return false;
+
+            try
+            {
+                return false;
+            }
+            catch (Exception ex)
+            {
+                LastException = ex;
+                LogError(ex.Source, ex.ToString());
+                throw;
+            }
+        }
```

Comment:
This is just returning false in all cases, and no exception could possibly occur.

```suggestion
        public override bool RefreshData()
        {
            return false;
        }
```